packages/commands/src/index.ts

```diff
@@ -615,8 +634,31 @@ export class CommandRegistry {
    * Execute the command for the given key binding.
    *
    * If the command is missing or disabled, a warning will be logged.
-   */
-  private _executeKeyBinding(binding: CommandRegistry.IKeyBinding): void {
+   *
+   * The execution will not proceed if any of the events leading to
+   * the keybinding matching were held with the permission resolving to false.
+   */
+  private async _executeKeyBinding(
+    binding: CommandRegistry.IKeyBinding
+  ): Promise<void> {
+    if (this._holdKeyBindingPromises.size !== 0) {
+      // Wait until all hold requests on execution are lifted.
+      const executionAllowed = (
+        await Promise.all(
+          this._keydownEvents.map(
+            async event =>
+              this._holdKeyBindingPromises.get(event) ?? Promise.resolve(true)
+          )
+        )
```

Comment:
Should we add a race condition with a timeout to avoid a dead lock?

Reply:
There is already a timeout in https://github.com/jupyterlab/jupyterlab/pull/15927. I will add a timeout here too.

Reply:
Done in c5c0a9c7084fd4ecf876053bfc982d6cdd3d76e8.